lib/utils/http.go

```diff
@@ -116,3 +116,19 @@ func GetAnyHeader(header http.Header, keys ...string) string {
 	}
 	return ""
 }
+
+// CloneRequest makes a clone of the request. Request body is returned as well.
+func CloneRequest(req *http.Request) (*http.Request, []byte, error) {
+	body, err := GetAndReplaceRequestBody(req)
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+
+	clone, err := http.NewRequestWithContext(req.Context(), req.Method, req.URL.String(), io.NopCloser(bytes.NewReader(body)))
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+	clone.Header = req.Header.Clone()
+
+	return clone, body, nil
+}
```

Comment:
I wonder if we shouldn't just use `req.Clone(ctx)` and handle copying the body?

This makes an outgoing request with the same context as the incoming request's context, but it may be better to use the context passed into `HandleConnection`. You could pass that in here though and just use it instead of the request's context.

Since these requests come from `http.ReadRequest(clientConn)`, I think the context is `context.Background()` which means no outgoing request cancellation. I may be wrong though, I just don't see `req.ctx` being set in that method.

Reply:
I looked at this; indeed, `req.Clone(ctx)` is much more common in our codebase. At the same time, there are certain assumptions made about request, which make it hard to use the `CloneRequest` correctly. I came to see the `CloneRequest` as unsuitable for shared usage as it is: after the request received from the client is cloned, we need to override `Host` and `RequestURI` fields to be able to issue the request ourselves, among other things. I don't believe singular function is a proper way to share this functionality, but rather we should go in the direction of shared middleware:

https://github.com/gravitational/teleport/blob/a6abd933647c01f311ccfafe8214d3005a250b03/lib/srv/db/opensearch/engine.go#L152-L154